WIP: joe/move p zigzag v2

meese-wj opened the pull request:
Different approach to zigzags.

rfbird committed:
Add note on GCC version to readme

rfbird committed:
add doc listing the level of support for different features

SVLuedtke-LANL committed:
Merge branch 'devel' into kokkos3
Upgrading Kokkos to version 3.

SVLuedtke-LANL committed:
Merge branch 'kokkos3' into devel
Upgrading to Kokkos 3.

rfbird committed:
add extra mpi includes for systems that can't find mpi in input decks

rfbird committed:
quick update to README to cover Kokkos3 cmake instructions

SVLuedtke-LANL committed:
Merge branch 'add_extra_mpi_includes' into devel

Passes MPI includes to decks and addresses #13.

SVLuedtke-LANL committed:
Semantics for the built-in dumps are now the same as in the non-Kokkos
version.  The dump functions themselves handle the device to host copy,
and will check if the data are already current.

Also provided are user_diagnostics_copy_*_mem_to_host functions for users to
easily copy data to the host for their diagnostics.  These will also
check if the data are already current.

meese-wj committed:
First commit on the zigzag branch

I updated species_advance.h to be built with either the zigzag
current accumulation or with the old Villasenor method by setting
ACCUMULATE_ZIGZAG when building with CMake. This is the first
commit before I start messing with current accumulation.

meese-wj committed:
Added first-order weighting to the currents

The current accumulator has been modified so reflect first-order
weighting. Initial tests of the Harris deck have seen no results
though.

meese-wj committed:
Fixed the CMake error

The cmake flag actually works now... add_definitions(-D[flag])
will default to true regardless, so pulling it inside the if
statement worked.

Commit:
Fixed the minor format/style issues

The more pressing issues regarding what dt (the size of the time
step) is and whether the l argument in accumulate_j_zigzag(...) is
actually necessary are currenlty left unresolved as they require
more thought. TODOs have been left in the comments as well.

rfbird committed:
Merge pull request #17 from lanl/dump_autocopy

Improved Dump Semantics

Commit:
Fixed a typo in Fz2

I originally copied it inline as Fx2. This didn't affect anything
before since l = 2 was never called in accumulate_j_zigzag(...).

Commit:
Zigzag Harris energy reduction by a factor of 100

This point is not closed but part of the issue with what I was
doing before was using the midpoints and normalized displacement
calculated after the comment: // Compute the midpoint and the
normalized displacment of the streak. I used them as if they were
the initial point and total displacement of the particle.

This does not solve all the problems, but it represents a start.
Currently, accumulate_j_zigzag.hpp is full of "joe_" variables so
as not to interfere with s_midx, s_dispx, etc., in the rest of
move_p_kokkos(...).

Commit:
In the process of debugging

And so printf statements are flying around like crazy. One thing
that I caught was the in the reference point calculation, I forgot
to add a constant 1.f inside the outermost fmin(...). This is
fixed and the reference point now appears to be correct. The
numerical values for the energies and fields are again within an
order of magnitude and usually within 10% of the standard move_p_
kokkos(...) implementation. I'm not yet sure where the final
discrepancy is from though.

Commit:
Still debugging the current accumulation

Something is weird regarding the in-cell coordinates compared
to the Umeda paper. For some reason current is not being
accumulated on all parts of the cell where it should be -- right
now it's only accumulated on the v0 edge. Also, for some reason
move_p_kokkos(...) is being entered when the particle does not
leave the cell.

The main changes right now are print statements outputting the
current accumulation values, and I reverted the changes in the
particle position and displacements for simplicity (they changes
are currently commented out).

Commit:
Cell and Umeda coordinates conversion macro-ed up

A quick hack resulting in some progress along the current
accumulation lines has been done by converting between cell
coordinates which run from [-1,1] and the coordinates assumed in
Umeda et al, "Umeda coordinates", [0,1]. The current using this
method along the zig of the zigzag is regularly between a factor
of 5-10 smaller than the older current accumulation method.
However the 2particle.cxx deck is almost identical through 500
time steps. This is may be due to the deck's simplicity.

Ignore giant swaths of commented-out code. That will be removed
once these macro hacks are Bob-certified and promoted to
officially-sanctioned code. (They are left in for the sake of
future discussion.)

It still needs to be verified that the "two zig" method for the
zigzag actually is a zigzag and not just wishful thinking.

Commit:
Trying something new next

The coordinate conversions, included those for the fluxes, seem
weird. I don't like the random factors of 8. Also, currents were
not being deposited at all on the zag, so I added quite a few if
statements to turn on l = 2.

At this point I'm going to try something different and run away
from the dumpster fire this code's becoming. If I can't figure it
out with the new idea with the weights, then come back to this
commit.

Commit:
First commit on the new zigzag branch

In this branch I'll keep the deposition almost identical to that
of the old method before and simply modify the particle motion
along the zig or the zag.

Commit:
Moved some things around and started zigzag again

Since it looks like the particle weight factors between the
Villasenor and Umeda algorithms are almost the same (up to that
extra 1/12 term in the Villasenor paper), the idea now is to move
the particles manually along the zigzag, and then have the
current accumulate over this new trajectory.

The end result is a lot of if statements that should be, in
principle, possible to cut down once we know the idea works. The
code runs and spits out the correct 1d zig currents, but fails on
the 1d zag currents. I haven't tried higher dimensions. I also
added more comments to the older code to help future readers
understand move_p_kokkos(...) without as much difficulty.

Two things that will need to be addressed. First, even in the old
code, it looks the the particle displacement is unnecessarily
multiplied by a factor of 2 when axis == 3. This I do not quite
get yet, and so I don't know how important it is. Second, it is
not clear how to shoot a particle through a corner instead of only
along one cube face as it is done now. If this cannot be easily
done, it's not clear that the zigzag algorithm will actually be
any different than what's currently implemented in the code.

Commit:
Zig and Zag seem to be working

The zigzag method works when s_disp is taken to be half of the
actual displacement that the particle is going to move along as
calculated by advance_p. Additionally, the displacements should
only be further halved if the particle leaves the cell -- that
is, when axis != 3. The 2particle deck accumulates the same
currents now between the two methods now.

The particles now must be properly moved to the boundaries such
that edge or corner crossings are allowed.

Commit:
Added documentation on displacements

I added some comments on how the displacments change throughout
the move_p_kokkos(...) function in both accumulate_j.hpp (the old
method) and in accumulate_j_zigzag.hpp (the zigzag method). The
goal is to make future development in this function easier as the
translation is already present.

rfbird committed:
start extending #neigbhors to be full 3d => 26

rfbird committed:
fix typo in variable name that was stopping compilation

Commit:
Merge branch 'extend_neighbor_array' into joe/move_p_zigzag_v2

Stealing Bob's extended neighboring scheme into my branch to
modify it into my own.

Commit:
Just an update for Kokkos

There are uncommitted changes to kokkos from Bob's branch. This
commit is just to separate my work from Kokkos.

Commit:
Changed up the local neighborhood for each cell

Now the simulation segfaults right before move_p_kokkos(...) is
called. I'm not sure where it is, but the simulation does not
actually even enter the move_p part of the conditional in
advance_p.cc. It's odd that this wasn't apparent yesterday...

Commit:
Weirdness with the segfault

The generalization from 6 * pii + face in move_p_kokkos(...) should
be 27 * pii + neighbor_index. My first change got it wrong as
just neighbor_index, and there was no segfault. As soon as I put in
the generalization, the simulation segfaults at time step 260 in the
2particle deck BEFORE entering move_p_kokkos(...) for the first
time. By taking out the 27 * pii + and just having neighbor_index,
the segfault is again removed.

Commit:
Quick commit before checking devel

See above.

Commit:
Fixed the segfault

In init_kokkos_grid(...) in initialize.cc on devel, the size of
the neighbor array was set as the input argument. After changing
the size of the local neighbors from 6 to 27, this input argument
was taken out but it needed to be 27 * number_of_voxels. This is
"fixed" now, and so I can continue figuring out why the energies
are off.

Commit:
Added some fixes and TODOs regarding the neighbors

There are also some debugging print statements. It looks like the
new neighbors have also changed the way that the old code (not
zigzag) works because the energies and current accumulations are
different.

Commit:
Added some more print statements

Tracking down a bug. Going to checkout an older commit and diff
it up.

Commit:
Zigzag and straight lines now agree while wrong

So now both implementations are calculating an energy that is
about 2x too small than before I messed with the neighbor array.
But they agree now, and so I think the next step is verifying the
neighbor array is not wrong.

Commit:
neighbor array seems to be mis-mapped

Between size_grid and init_kokkos_grid, the values of the neighbor
array change from being "correct" to crazy (a cell in the middle
of the simulation only connects to itself or reflect_particles).

Commit:
The problems came from GLUE_FACE!

GLUE_FACE needed to be adjusted for the change in the labelling
scheme for the cube faces. I also added an inline function for
computing the neighbor indices (get_neighbor_index) inside of
grid.h. The constexpr was ditched since we don't always have
the arguments at compile time (looking at you move_p_kokkos).

Commit:
Added an Axis_Label enumeration for readability

This is simply to enumerate indices in move_p_kokkos to help with
the readability in the code.

Commit:
Starting to dump the neighbor indices

I modified the CMakeLists.txt in the parent directory to add a
VPIC_DUMP_NEIGHBORS flag for the build. This will print the value
of each neighbor index for future comparison.

I will now add a macro to print the indices if the build flag is
ON.

Commit:
Added a functor to dump neighboring indices into

I have to change a couple of things, then it can be used to
plot histograms of the neighboring indices a particle was sent
through in VPIC.

Commit:
Removed some comments and finished the functor

Now the neighbor indices can be plotted as histograms which show
the number of instances a particle is moved to a particular
neighbor with index 0, 1, 2, ..., 26.

Additionally, there is no discrepancy in the energies.txt files
for the harris deck between devel and the Villasenor method with
27 neighbors on this branch. Currently zigzag does not agree with
these methods. Interestingly enough, there are more neighbor
indices for zigzag than there are for Villasenor implying that
there are more move attempts in zigzag. This seems wrong...

Commit:
Added an image of the neighbor indices in ops.cc

I also removed some commented out older attempts at extending the
local neighborhood of each voxel from 6 neighbors to 26 and the
voxel itself.

Commit:
Added/Subtracted print statements for debugging

It looks like GLUE_FACE and SET_PBC in ops.cc must be modified for
the new neighborhood. Not sure how deep this bug is, but I might
as well fix it now because it affects my 2particle_diagonal tests.

Commit:
Changed BOUNDARY to be consistent

Now it calculates the neighbors in the same way as the inline
get_neighbor_index. I'm surprised VPIC didn't bug out before
because of this. Maybe it's because SET_PBC has a seemingly
unnecessary IF statement in it?

rfbird reviewed (changes requested):
Looks mostly sensible! Great start

Commit:
Added another filestream to print planes

These planes take on one of the following values {-1, 0, 1} along
each axis. This is used to plot whether the neighbor a particle
moves into is a Face, Edge, or Corner.

Commit:
Added final destination connection writer

This poor VPIC_DUMP_NEIGHBORS "functor" is really more of just an
output stream object at this point. Anyway it now determines if
each particle moves through a Face, Edge, or Corner and writes
that info to a file for future analysis. It's smart enough to only
do this once per particle to avoid over-counting. This can be used
to estimate what the likely speedup zigzag will have over the
straight-line method which can only send particles through faces.

Commit:
Found a bug in zigzag. Going to work on it now

Zigzag does not find the midpoint along an axis along which the
particle does not change cells.

Commit:
Update on zigzag bug

The bug was not that the midpoint was wrong, it was that the
current accumulation was not calculated at the "quarter-point", as
expected in the accumulate_j macro. Instead what was happening is
the currents were accumulated relative to the full displacement
rather than the displacement to the midpoint first along the zig
before proceeding to the zag.

Commit:
Previous bug fixed I think. Need tests

I was able to get around the problem by introducing three new
floats: finalx, finaly, and finalz. These tell the particle where
to end up and I think now I can collapse some extra IF statements
(assuming this approach is correct).

Commit:
Displacements need to be inside IFs

These are the same IFs that I'm pretty sure I can eliminate by
merging them with ones that come before. So far it looks like the
particle positions are what they should be and the midpoints along
the zig and zag (so-called "quarter-points") are used to calculate
the current accumulations. So things look promising right now!

There are still systematic discrepancies that are expected for
large numbers of particles moving through edges and corners. A more
thorough analysis is needed to test the veracity/fidelity of VPIC
with "working" zigzags.

Commit:
Zigzag seems to be working; Harris disagrees

The particle positions seem to be correct and the currents then
seem to follow. Not sure why the Harris energies are so different
though.

Changed a comment for clarity.

Commit:
Harris agrees very wel for a single time step

After the first time step, the particle energies in the Harris
deck seem to be different only in the 5 significant figure. This
is heartening, as it means that the problems again may be due to
the lack of periodic boundary conditions with extended neighbors.

For another test, I will now change advance_p so eliminate any
contribution from the fields to study the particle motion.

Commit:
Cleaned up zigzag code a bit for readability

It's still hard to tell the degree of wrongness associated with
zigzags. It looks like the particles are moving along the proper
paths and depositing current appropriately. There is still a global
periodic boundary condition problem thanks to GLUE_FACE and SET_PBC,
but I'm not so sure how relevant they are.

Commit:
Commented out a bunch of prints

This code will be compared to devel in performance tests on
dedicated nodes.

Commit:
Added a couple files for performance analysis

Just a bash script and a couple of decks.

Commit:
Added some automated timing scripts

These will parse through VPIC's stderr for the Per column of
advance_p timings.
